pkg/apiserver/rest/dualwriter_mode2.go

```diff
@@ -0,0 +1,207 @@
+package rest
+
+import (
+	"context"
+	"fmt"
+
+	apierrors "k8s.io/apimachinery/pkg/api/errors"
+	"k8s.io/apimachinery/pkg/api/meta"
+	metainternalversion "k8s.io/apimachinery/pkg/apis/meta/internalversion"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/runtime"
+	"k8s.io/apiserver/pkg/registry/rest"
+	"k8s.io/klog"
+)
+
+type DualWriterMode2 struct {
+	DualWriter
+}
+
+// NewDualWriterMode2 returns a new DualWriter in mode 2.
+func NewDualWriterMode2(legacy LegacyStorage, storage Storage) *DualWriterMode2 {
+	return &DualWriterMode2{*newDualWriter(legacy, storage)}
+}
+
+// Get overrides the default behavior of the Storage and retrieves an object from
+// LegacyStorage or Storage depending on the DualWriter mode.
+func (d *DualWriterMode2) Get(ctx context.Context, name string, options *metav1.GetOptions) (runtime.Object, error) {
+	legacy, ok := d.legacy.(rest.Getter)
+	if !ok {
+		return nil, fmt.Errorf("legacy storage rest.Getter is missing")
+	}
+
+	l, err := legacy.Get(ctx, name, &metav1.GetOptions{})
+	if err != nil {
+		return nil, err
+	}
+
+	s, err := d.Storage.Get(ctx, name, &metav1.GetOptions{})
+	if err != nil {
+		return l, nil
+	}
+
+	accessorL, err := meta.Accessor(l)
+	if err != nil {
+		return l, err
+	}
+
+	accessorS, err := meta.Accessor(s)
+	if err != nil {
+		return l, err
+	}
+
+	accessorL.SetResourceVersion(accessorS.GetResourceVersion())
+	return l, nil
+}
+
+// List overrides the default behavior of the Storage and retrieves objects from
+// LegacyStorage or Storage depending on the DualWriter mode.
+func (d *DualWriterMode2) List(ctx context.Context, options *metainternalversion.ListOptions) (runtime.Object, error) {
+	legacy, ok := d.legacy.(rest.Lister)
+	if !ok {
+		return nil, fmt.Errorf("legacy storage rest.Lister is missing")
+	}
+
+	legacyList, err := legacy.List(ctx, options)
+	if err != nil {
+		return nil, err
+	}
+	ll, err := meta.ExtractList(legacyList)
+	if err != nil {
+		return nil, err
+	}
+
+	storageList, err := d.Storage.List(ctx, options)
+	if err != nil {
+		return nil, err
+	}
+	sl, err := meta.ExtractList(storageList)
+	if err != nil {
+		return nil, err
+	}
+
+	type fieldsToEnrich struct {
+		resourceVersion string
+	}
+
+	m := map[string]fieldsToEnrich{}
+	for _, obj := range sl {
+		accessor, err := meta.Accessor(obj)
+		if err != nil {
+			return nil, err
+		}
+		m[accessor.GetName()] = fieldsToEnrich{accessor.GetResourceVersion()}
```

Comment:
hm, should we have an error check here in case the name isn't present? Otherwise, it might panic here.

Reply:
Hmm good question. I wonder if it's ever possible to not have a name. Have to look into it.